### DISCUSSION TITLE
Reading constant names from constant nodes

### DISCUSSION THREAD

**Discussion by exterm** [ruby/prism#2434] [Q&A] [answered]
Hi,

I need to get the constant name from the different types of constant nodes:

ConstantWriteNode, ConstantPathNode, ConstantReadNode, ConstantPathWriteNode.

It's different for all of them, but there seem to be some especially strange inconsistencies.

For example, why does `ConstantReadNode` have a `full_name` method, although it's not a path node? `ConstantWriteNode` doesn't have that method.

Why does `full_name` return a string, but `name` returns a symbol?

**Comment by exterm** [comment]
cc @kddnewton (not sure how notifications for discussions work by default)

**Comment by kddnewton** [comment]
The `node_ext.rb` stuff is meant to provide convenience APIs if people need it. The `ConstantReadNode` has `full_name` in order to be compatible, so that you can call `full_name` on any kind of `Constant*` node and get back the same kind of stuff. `ConstantWriteNode` doesn't have it because you're always writing to a single constant name, but I'd be happy to have it.

`full_name` returns a string because it will include `::`, `name` returns a symbol because it's always a single identifier.

**Reply by exterm** [reply]
> The `ConstantReadNode` has full_name in order to be compatible, so that you can call full_name on any kind of `Constant*` node and get back the same kind of stuff. `ConstantWriteNode` doesn't have it because you're always writing to a single constant name, but I'd be happy to have it.

These two sentences seem contradictory. It'd be great to have `full_name` on `ConstantWriteNode` too so that we can call `full_name` on any kind of `Constant*` node.

**Reply by exterm** [reply]
I understand the way the code is generated from the config this is not currently really possible, but it would be _really_ nice to have some inheritance between node types.

`ClassNode` and `ModuleNode` share an interface, and all the `Constant*` nodes (could) also share one. For use with sorbet, it would be great to just be able to say `params(node: Prism::ConstantNode)`.

**Reply by kddnewton** [reply, marked as answer]
> These two sentences seem contradictory. It'd be great to have full_name on ConstantWriteNode too so that we can call full_name on any kind of Constant* node.

@exterm yeah as mentioned, I'd be happy to support `full_name` on `ConstantWriteNode`. Do you want to open a PR for that?

> it would be really nice to have some inheritance between node types.

I would be amenable to marker interfaces, which should work with Sorbet. That's been discussed a bit here: https://github.com/ruby/prism/discussions/2092 and I think this is something we're going to implement sometime this year.

**Reply by exterm** [reply]
PR for `full_name` on `ConstantWriteNode`: https://github.com/ruby/prism/pull/2485

**Comment by kddnewton** [comment]
I'm not sure if I answered all of your questions, feel free to ask more.
